rrs.1
.TH rrs 1 "7 December 2018" "version 1.0.1"
.SH NAME
rrs - RSSI Reader Software
.SH SYNOPSIS
.B rrs -h
.br
.B rrs --help
.br
.B rrs -V
.br
.B rrs --version
.br
.B rrs  [-c 
.I count
.B ]
.B  [-s 
.I seconds
.B ]
.B  [-m
.I milliseconds
.B ]
.B  [-i 
.I milliseconds
.B ]
.B  [-t 
.I option
.B ]
.B  [--noquiet]
.B  [--nooverwrite]
.B  [--noheaders]
.B  [-d 
.I delimiter
.B ]
.B  [--notransmit]
.B  [-p 
.I size
.B ]
.B  [-f 
.I file
.B ]
.SH DESCRIPTION
.B rrs
is a (fairly) highspeed IPv4 utility that reads a wireless network card's
connection strength values. The Radio Signal Strength Indication values are
read up to 10 times per second and saved to an output file (and optionally
displayed to the console).
.B rrs
requires an established wireless network connection. The RSSI value is updated
by the wireless nic when a packet is recieved from the wireless access point
or router. To ensure consistent RSSI updates
.B rrs
[optionally] transmits ICMP ECHO_REQUEST datagrams via the
.B ping
utility.

." disclaimer paragraph
.PP
Note that while this utility should work on other versions of linux, it has only
been tested on Raspian 9 (Stretch) on a Raspberry Pi 3 B+ with the onboard
wireless nic and Ubuntu 18.04 (Bionic Beaver) using a Belkin USB wireless nic.
Each device manufacturer's implementation of RSSI reporting may vary, therefore
the format and meaning of recorded values should be treated as inconsistent
across devies.
.B rrs
polls the procfs file /proc/net/wireless for RSSI data. The implementation of
procfs can vary between distros, which may render
.B rrs
incompatible with certain linux flavors.

.SH EXAMPLES
." add some examples
.B rrs -h
.RS
Print all commandline options and their default values to the console.
.RE
.B rrs
.RS
Run rrs with defaul values to produce a logfile of RSSI values.
.RE
.B rrs -i 100 -c 10000 --noquiet | sudo feedgnuplot --stream .1 --xlen 100 
.B "    --ymin -90 --ymax -10 --lines --xlabel 'Time' --ylabel 'RSSI'"
.RS
This example must be run in an X-Windows environment and requires feedgnuplot
(https://github.com/dkogan/feedgnuplot). The ymin and ymax values here are for 
Raspberry Pi 3 B+, adjust to your readings as necessary. The resulting output is
a scrolling line graph of real-time RSSI value changes.
.RE

.SH FILES
.I ./lib/
.RS
Support files needed for
.B rrs
to work.

.SH OPTIONS
.TP
.BI "-c, --count " count
Record by count. Specifies the number of RSSI readings to take. This option takes precedence over
.B --seconds
or
.BR --milliseconds .
If
.B --count 
is specified, record by time options are ignored.
.TP
.BI "-s, --seconds " seconds
Record by time. Specifies how many seconds to record RSSI readings for. This option combines with
.B --milliseconds 
additively. If
.B --count
is specified, record by time options are ignored.
.TP
.BI "-m, --milliseconds " milliseconds
Record by time. Specifies how many milliseconds to record RSSI readings for. This option combines with
.B --seconds 
additively. If
.B --count
is specified, record by time options are ignored. The fastest RSSI values can be read is 100 ms.
Any value lower than that is treated as 100 ms. 
.TP
.BI "-i, --interval " milliseconds
Specifies the frequency of RSSI readings to take given in milliseconds. The fastest RSSI values can be
read is 100 ms. Any value lower than that is treated as 100 ms.
.TP
.BI "-t, --timestamp " OPTION
Specifies the format of the timestamp that prepends each RSSI reading.  Timestamps increment roughly
equal to the
.B --interval
and is measured in milliseconds.
.br
    zero - Uses a timer starting at 0.
    now  - Prints the local time for each reading.
    none - No timestamp is added.
.TP
.B "-q, --[no]quiet "
Suppress readings output to stdout. To have the reading displayed to the console as well as written to
the log file use either 
.BR --noquiet " or " --quiet
.IR false .
.TP
.B "-o, --[no]overwrite "
Overwrite existing log file if it has the same file name as specified (or default). If disabled log
files will be appended to. To disable use
.BR --nooverwrite " or " --overwrite
.IR false .
.TP
.B "-H, --[no]headers "
Write column headers as the first line written to the log file. To disable writing headers use
.BR --noheaders " or " --headers
.IR false .
.TP
.BI "-d, --delimiter " String
Defines the column separator for each RSSI reading output. If the delimiter includes whitespace it must
be surrounded by double quotes.
.TP
.B "-x, --[no]transmit "
Transmit ICMP packets to the access point. To disable use
.BR --notransmit " or " --transmit
.IR false .
.TP
.BI "-p, --packetsize " Integer
Specifies the padding size of transmitted ICMP datagram packets. The default is 8,
which translates into 16 ICMP data bytes when combined with the 8 bytes of the 
ICMP header data.
.TP
.BI "-f, --file " file
Specifies the path and filename that RSSI values will be recorded to.
.TP
.B "-V, --version "
Prints the current version and copyright statement.

.SH BUGS
." do we have any bugs or just features?
.B Known Bugs
.RS
The delimiter field can not process escape characters and will cause the program
to fail. The wireless device name is currently hardcoded to wlan0. To change this
edit the rrs script file.
.RE

.SH AUTHORS
." that's me
Ivan Boatwright & Julieth Diaz

.SH COPYRIGHT
Copyright (c) 2018 Ivan Boatwright

rrs comes with NO WARRANTY, to the extent permitted by law.
For information about the terms of redistribution,
see the file named README in the rrs distribution.

rrs.sh
#!/usr/bin/env /bin/bash

#command line parameters: "number of packets to send", "scan interval"
# then loop doing a scan each pass while the packets are being sent... 
#     whether using ping or directly injecting packets, I dont care...
# the output from the scan is appended to the logfile with a timestamp1

# source shflags
. ./lib/shflags

# DEFINE_string 'dev' 'wlan0' 'device to read RSSI values from' 'd' 



######### UserInterface ############
####################################
# configure shflags
DEFINE_integer 'count' 100 'number of RSSI values to record' 'c'
DEFINE_integer 'seconds' 0 'number of seconds to record' 's'
DEFINE_integer 'msecs' 0 'number of milliseconds to record' 'm'
DEFINE_integer 'interval' 500 'frequency in milliseconds to record RSSI values' 'i'
DEFINE_string 'timestamp' 'zero' 'logged timestamp format: (zero | now | none )' 't'
DEFINE_boolean 'quiet' true 'suppress message output to stdout' 'q'
DEFINE_boolean 'overwrite' false 'overwrite old log files' 'o'
DEFINE_boolean 'headers' true 'prepend logfile with column headers' 'H'
DEFINE_boolean 'transmit' false 'transmit packets to the access point' 'x'
DEFINE_integer 'packetsize' 8 'size of the packet to transmit +8 bytes for header' 'p'
DEFINE_string 'file' 'log.csv' 'file to save the RSSI log results' 'f'


# parse the command-line
FLAGS "$@" || exit $?
eval set -- "${FLAGS_ARGV}"

# assigns input from the cli to the corresponding flag variable
setFlags()
{
	# sets the delay between readings to seconds (with fractions)
	RSSI_READ_INTERVAL=$(echo "scale=4; $FLAGS_interval/1000" | bc -l)"s"

	# sets the proper timestamp to prepend to each reading
	[[ $FLAGS_timestamp -eq "zero" ]] && START_TIME=$(date +%s%3N) || START_TIME=0
	DELTA=3 # small offset to adjust for START_TIME recording to first timestamp

	# toggles whether to print output to the screen as well as the log file
	[[ $FLAGS_quiet -eq $FLAGS_TRUE ]] && setDisplayPath "/dev/null" ||	setDisplayPath "/dev/fd/3"
}

# set the (nonpersistant) display path
setDisplayPath()
{
	displayPath=$1
}

# return the path being used for the display
getDisplayPath()
{
	echo "$displayPath"
}

# verify the user input is valid or exits
validateInput()
{
	[[ $1 -ne $2 ]] && echo "Invalid input format." >&2;
	exit 1;
}

# print the commandline options to the screen
printOptions()
{
	if [[ -e /dev/fd/3 ]]; then
		rrs.sh -h >&1;
	else
		rrs.sh -h | tee /dev/fd/3
	fi
}

# display pretty print error messages
printErrors()
{
	local errMsg
	[[ -z "$1" ]] && errMsg="An error has occurred. RRS is terminating." || errMsg="$1"
	if [[ -e /dev/fd/3 ]]; then
		echo "$errMsg" | tee /dev/fd/3;
		exit 1;
	else
		echo "$errMsg";
		exit 1;
	fi
}

########### PacketTransmitter #############
###########################################
# Sends packets to the access point
sendPacket()
{
	ping -4nq -c $packetCount -i $transmitInterval -I getDevName -s packetsize getAPIP &
	pingPID=$!
}

# Cancels the active packet transmission
cancelTransmission()
{
	kill -SIGTERM $pingPID;
}

# Returns the name of the wireless network adapter (grabs the first in case of
#  multiple adapters.)
getDevName()
{
	echo "iw dev | awk '/Interface/ { print $2 }'"
}

# Gets the access point's IP address
getAPIP()
{
	local var=$(getDevName)
	echo $(ip route show | grep $var | awk 'NR==1 { print $3 }')
}

# Init function for the PacketTransmitter 'class'
packetTransmitter()
{
	packetCount=$FLAGS_count
	transmitInterval=$FLAGS_interval
	packetsize=$FLAGS_packetsize
	destination=getAPIP
}
########### RSSIParser #############
####################################
# This parser is specific to Raspbian 9 (stretch), the value reported for quality
#		level is determined by the wireless nic's driver. If you want any other
#		values reported in /proc/net/wireless just add their respective capture
#		class ($3 should be quality link and $5 quality noise).
parseRSSIValue()
{
	#echo $(getDisplayPath)
	if [[ $1 -eq true ]]; then
		awk -F ".[ ]+" -v date="$(($(date +%s%3N)-START_TIME-DELTA))" 'NR==3 {print date "," $4}'	/proc/net/wireless | tee $(getDisplayPath)
	else
		awk -F ".[ ]+" 'NR==3 {print $4}' /proc/net/wireless | tee $(getDisplayPath)
	fi
}

# Calculates the time offset from user input
calculateEndSeconds()
{
	END_SECONDS=$(echo "scale=4; $FLAGS_msecs/1000 + $FLAGS_seconds" | bc -1)
}

# Calculates the future time to stop reading RSSI values
calculateStopTime()
{
	STOP_TIME=$(date +%s%3N -d "+$END_SECONDS")
}

##
#  Read the RSSI value for a set number of times
#
readRSSIByCounter()
{
	if [ ${FLAGS_timestamp} != "none" ]; then
		for i in `seq 1 ${FLAGS_count}`;
		do
			parseRSSIValue true
			DELTA=0
			sleep $RSSI_READ_INTERVAL
		done
	else
		for i in `seq 1 ${FLAGS_count}`;
		do
			parseRSSIValue false
			sleep $RSSI_READ_INTERVAL
		done
	fi
}

##
#	 This parser repeats until the current date is greater than or equal to the stop time.
#
readRSSIByTimer()
{
	calculateEndSeconds
	calculateStopTime	
	if [ ${FLAGS_timestamp} != "none" ]; then
		for i in `seq 1 ${FLAGS_count}`; # while (date -lt STOP_TIME);
		do
			awk -F ".[ ]+" -v date="$(($(date +%s%3N)-START_TIME))" 'NR==3 {print date "," $4}'	/proc/net/wireless | writeToFile $(getDisplayPath)
			sleep $RSSI_READ_INTERVAL
		done
	else
		for i in `seq 1 ${FLAGS_count}`; # while (date -lt STOP_TIME);
		do
			awk -F ".[ ]+" 'NR==3 {print $4}' /proc/net/wireless | writeToDisplay $(getDisplayPath)
			sleep $RSSI_READ_INTERVAL
		done
	fi
}


########### RSSILogger #############
####################################
# Clear the contents of the target log file
overwriteFile()
{
	rm $FLAGS_file
	touch $FLAGS_file
}

# Verify the output file exists
verifyOutput()
{
	[[ -e "$FLAGS_file" ]] && echo true
}

# compute a timestamp for logging
calculateTimeStamp()
{
	timeStamp="$(($(date +%s%3N)-START_TIME))"
}

# get the last computed timestamp value
getTimeStamp()
{
	echo "$timeStamp"
}

# redirect the value passed to this function to the display
writeToDisplay()
{
	[[ true -eq $((0)) ]] && tee $1
}

# redirect the value passed to this function to the file
writeToFile()
{
	[[ $((0)) -eq true ]] && tee $1
}

##
# Handles creating & preparing new log files if needed. Also tests for write access.
##
initLogFile()
{
	if [ -w "$FLAGS_file" ]; then
		if [ $FLAGS_overwrite -eq $FLAGS_TRUE ]; then
			overwriteFile
		fi
	elif [ -e "$FLAGS_file" ]; then
		printErrors "Error! Cannot overwrite the logfile $FLAGS_file."
		exit 1;
	else
		touch $FLAGS_file
	fi

	# set all 1 & 2 output to append to the file while leaving fd 3 attached to stdout
  exec 3>&1 1>>${FLAGS_file}
  [[ $FLAGS_headers -eq $FLAGS_TRUE ]] && echo "time,RSSI"
}


########
# Main #
########
setFlags
initLogFile

#[[ $FLAGS_transmit -eq true ]] && packetTransmitter && sendPacket
# Counter or Timer?
if [ $((FLAGS_seconds + FLAGS_msecs)) -gt 0 ]; then
	readRSSIByTimer
else
	readRSSIByCounter
fi

###### old commands I might need ######
#sudo iwlist wlxec1a595e181d scan | grep 'Signal level=' | awk -F'[=]' '{print $3}' | head -n1
#watch -n 1 "awk 'NR==3 {print \"WiFi Signal Strength = \" \$3 \"00 %\"}''' /proc/net/wireless"